File: DanceCalc/m4dModels/UserQuery.cs
using System;
using System.Linq;
using System.Text;

namespace m4dModels
{
    // normalize to (+|-)UserName\|[l|h]
    public class UserQuery
    {
        public UserQuery(string query = null)
        {
            Query = Normalize(query);
        }


        public UserQuery(string userName, bool include, bool? like)
        {
            var qs = Normalize(userName);
            if (!string.IsNullOrEmpty(qs) && !string.IsNullOrWhiteSpace(qs) && qs != "null")
            {
                if (!include)
                {
                    qs = "-" + qs.Substring(1);
                }
                if (like.HasValue)
                {
                    if (like.Value)
                    {
                        qs += 'l';
                    }
                    else
                    {
                        qs += 'h';
                    }
                }
            }
            Query = qs;
        }

        public UserQuery(UserQuery query, string userName) : this(userName, query.IsInclude, query.NullableLike)
        {
            var qs = Normalize(userName);
            if (!query.IsEmpty && !string.IsNullOrWhiteSpace(qs) && qs != "null")
            {
                if (query.IsExclude)
                {
                    qs = "-" + qs.Substring(1);
                }
                if (query.IsLike)
                {
                    qs += 'l';
                }
                else if (query.IsHate)
                {
                    qs += 'h';
                }
            }
            Query = qs;
        }

        private bool? NullableLike
        {
            get
            {
                if (IsLike) return true;
                if (IsHate) return false;
                return null;
            }
        }

        private static string Normalize(string query)
        {
            if (string.IsNullOrWhiteSpace(query))
            {
                return string.Empty;
            }
            if (string.Equals(query, "null", StringComparison.OrdinalIgnoreCase))
            {
                return "null";
            }

            query = query.Trim().ToLower();
            if (query[0] != '+' && query[0] != '-')
            {
                query = "+" + query;
            }
            if (!query.Contains('|'))
            {
                query += '|';
            }
            return query;
        }

        public string Query { get; }
        public bool IsNull => string.Equals(Query, "null", StringComparison.OrdinalIgnoreCase);
        public bool IsEmpty => string.IsNullOrWhiteSpace(Query) || IsNull;
        public bool IsInclude => !IsEmpty && Query[0] == '+';
        public bool IsExclude => !IsEmpty && Query[0] == '-';
        public bool HasOpinion => !IsEmpty && !Query.EndsWith("|");
        public bool IsLike => Query.EndsWith("|l") || Query.EndsWith("|L");
        public bool IsHate => Query.EndsWith("|h") || Query.EndsWith("|H");
        public string UserName => IsEmpty ? null : Query.Substring(1, Query.IndexOf('|') - 1);
        public bool IsAnonymous => string.Equals(AnonymousUser, UserName,StringComparison.OrdinalIgnoreCase);

        public const string AnonymousUser = "me";

        public string ActionDescription
        {
            get
            {
                if (IsNull) return "Don't filter on my activity";
                var start = IsInclude ? "Include only" : "Exclude all";
                string end;
                if (IsLike)
                {
                    end = "I marked LIKE";
                }
                else if (IsHate)
                {
                    end = "I marked DON'T LIKE";
                }
                else
                {
                    end = "I tagged";
                }

                return $"{start} songs {end}";
            }
        }

        public string ODataFilter
        {
            get
            {
                if (IsEmpty || IsAnonymous) return null;

                var inc = "any";
                var cmp = "eq";
                if (IsExclude)
                {
                    inc = "all";
                    cmp = "ne";
                }

                var like = NullableLike;
                var userName = UserName.ToLower();

                if (like.HasValue)
                {
                    return MakeOneOdata(userName, inc, cmp, NullableLike);
                }

                var a = MakeOneOdata(userName, inc, cmp, null);
                var b = MakeOneOdata(userName, inc, cmp, true);
                var c = MakeOneOdata(userName, inc, cmp, false);

                return IsInclude ? $"({a} or {b} or {c})" : $"!({a} and {b} and {c})";
            }
        }

        private static string MakeOneOdata(string userName, string inc, string cmp, bool? like)
        {
            var vote = string.Empty;
            if (like.HasValue)
            {
                vote = like.Value ? "|l" : "|h";
            }

            return $"Users/{inc}(t: t {cmp} '{userName}{vote}')";
        }

        public string Description(bool trivial = false)
        {
            if (IsEmpty) return string.Empty;

            string start;
            if (trivial)
            {
                start = IsInclude ? string.Empty : " not";
            }
            else
            {
                start = IsInclude ? " including songs" : " excluding songs";
            }
            var ret = new StringBuilder(start);

            if (IsLike)
            {
                ret.Append(" liked");
            }
            else if (IsHate)
            {
                ret.Append(" disliked");
            }
            else
            {
                ret.Append(" edited");
            }
            ret.Append(" by ");

            ret.Append(UserName);

            return ret.ToString();
        }

        public override string ToString()
        {
            return Description();
        }
    }
}

File: m4d/Client/src/model/UserQuery.ts
export class UserQuery {
  public static fromParts(parts?: string): UserQuery {
    if (!parts || parts === "NT") {
      return new UserQuery("");
    }

    if (parts.length !== 2) {
      throw new Error("parts should be exactly 2 characters");
    }

    let prefix = "";
    if (parts[0] === "I") {
      prefix = "+";
    } else if (parts[0] === "X") {
      prefix = "-";
    } else {
      throw new Error('1st character of parts must be "I" or "X"');
    }

    let suffix = "";
    switch (parts[1]) {
      case "L":
        suffix = "l";
        break;
      case "H":
        suffix = "h";
        break;
      case "T":
        suffix = "a";
        break;
      default:
        throw new Error(
          '2nd character of parts must be one of "L", "H", or "T"'
        );
    }

    return new UserQuery(prefix + "me|" + suffix);
  }

  private data: string;

  public constructor(query?: string) {
    this.data = this.normalize(query);
  }

  public get query(): string {
    return this.data;
  }

  public get isEmpty(): boolean {
    return !this.data;
  }

  public get include(): boolean {
    return this.data[0] === "+";
  }

  public get like(): boolean {
    return this.parts[1] === "L";
  }

  public get hate(): boolean {
    return this.parts[1] === "H";
  }

  public get tag(): boolean {
    return this.parts[1] === "T";
  }

  public get userName(): string {
    const idx = this.data.indexOf("|");
    return this.data.substring(1, idx);
  }

  public isDefault(userName?: string): boolean {
    if (!userName && !this.data) {
      return true;
    }
    const field = this.data?.toLowerCase();
    return field
      ? field === "-me|h" || field === `-${userName?.toLowerCase()}|h`
      : !field;
  }

  public get description(): string {
    if (this.isEmpty) {
      return "";
    }

    const include = this.include ? "including" : "excluding";
    let like = "";
    switch (this.parts[1]) {
      case "L":
        like = "liked";
        break;
      case "H":
        like = "disliked";
        break;
      case "T":
        like = "edited";
        break;
    }

    return `${include} songs ${like} by ${this.userName}`;
  }

  public get parts(): string {
    if (!this.data) {
      return "NT";
    }

    let ret = this.data[0] === "-" ? "X" : "I";

    const parts = this.data.split("|");
    if (parts.length > 1) {
      switch (parts[1]) {
        case "l":
          ret += "L";
          break;
        case "h":
          ret += "H";
          break;
        default:
          ret += "T";
      }
    } else {
      ret += "T";
    }

    return ret;
  }

  private normalize(query?: string): string {
    if (!query) {
      return "";
    }
    query = query.trim().toLowerCase();
    if (!query) {
      return "";
    }
    if (query === "null") {
      return query;
    }

    if (query[0] !== "+" && query[0] !== "-") {
      query = "+" + query;
    }
    if (query.indexOf("|") === -1) {
      query += "|";
    }
    return query;
  }
}
